docs/internal/foreign_code.md

```diff
@@ -62,32 +65,210 @@ We hope that the region abstraction will make programming with sandboxes natural
+Verona allows overloading most of the operators that C++ provides and so these can be trivially mapped just like any other methods.
+The main exceptions relate to assignment, though Verona's `apply` sugar provides similar functionality.
+Classes from C++ need to opt out of `create` and `apply` sugar because these reflect patterns that don't exist in C++ codebases and so it's possible that apply sugar could be applied for C++ assignment operators.
+
+### Casts
+
+C++ casts are exposed as generic functions on the class that represents the library.
+Any C library has a `$cast[T,U](U) : T` function defined, which takes any C built-in type or pointer type and returns any other.
+For C++ codebases, this performs a 'C-style cast', including invoking user-defined conversion operators.
+Any C++ library will also expose `static_cast[T,U](U) : T` and `reinterpret_cast[T,U](U) : T` functions and, if compiled with RTTI enabled, a `dynamic_cast[T,U](U) : T` function.
+These can all be used with any combination of types for which their C/C++ analogues are defined and with the same semantics.
+The Verona syntax implicitly applies to pointers when given any value except a C built-in type.
+C built-in types are passed as values to the cast functions unless wrapped in `Ref` cells.
+For example:
 
 ```verona
-var llvm = LLVM.Create(PROCESS);
+// Equivalent to auto *x = new SomeStruct();
+var x = C.SomeStruct();
+// Equivalent to auto *y = static_cast<SomeOtherStruct*>(x);
+var y = C.static_cast[C.SomeOtherStruct](x);
+var i : C.int = 12;
+// Equivalent to f = static_cast<float>(i);
+var f = C.static_cast[C.float](i);
+// Equivalent to float *f_type_punned = reinterpret_cast<float*>(&i);
+var f_type_punned : C.float = C.reinterpret_cast[Ref[C.float], Ref[C.int]](i);
 ```
 
-This would create a new process that loaded the LLVM library and had RPC invocations for any of the LLVM functions that this program called.
-The `llvm` object in this example is a new isolated region.
-The programmer can then construct LLVM types within that region.
+Note that the last line of this is explicitly violating type safety.
+All of these operations are restricted to values that are within a sandboxed foreign region where type safety is explicitly not guaranteed.
+
+### Exceptions
+
+**This doesn't make me very happy, better suggestions welcome**
+Every sandboxed function may throw a `SandboxError` exception if the sandbox has failed and so all invocations of sandboxed code must be wrapped in a `try` block.
+
+C++ has unchecked exceptions and so any function not explicitly declared as `noexcept` may throw any type.
+Verona, in contrast, has only checked exceptions.
+This provides us with a way of identifying the expected exceptions:
+Any exception of a type from a particular sandboxed library that the code invoking the foreign code either explicitly catches or advertises that it may throw is assumed to potentially originate in the sandbox.
```

Comment:
It may be hard to know which exceptions can be thrown, with standard exceptions being thrown from standard library functions (like `bad_alloc`).

Reply:
Right, and without checked exceptions a C++ function will happily propagate any exceptions that are thrown.  This is much worse in the general case with the Itanium ABI because a C++ function may call a gcj / Ada / whatever function that throws an exception that the C++ program doesn't know how to handle.  These can always be caught with `catch(...)` though, which means that we can always wrap calls in that and propagate the fact that an exception occurred (even if we don't know what the exception was) to Verona.

Reply:
I think this probably needs an example.